File: src/bnbprob/slpfssp/sigma.py
import copy

from bnbprob.slpfssp.job import Job


class Sigma:
    """Base sequence of jobs and partial completion times for machines"""

    jobs: list[Job]
    """Jobs included into sequence (permutation)"""
    C: list[list[int]]
    """Partial completion times for machines of each parallel semiline"""
    m: list[int]
    """Number of machines on each parallel semiline"""

    def __init__(self, jobs: list[Job], C: list[list[int]]):
        """Base sequence of jobs and partial completion times for machines

        Parameters
        ----------
        jobs : list[Job]
            Jobs included into sequence

        C : list[list[int]]
            Partial completion times
        """
        self.jobs = jobs
        self.C = C
        self.m = [len(c) for c in C]

    def __del__(self) -> None:
        self.cleanup()

    def cleanup(self) -> None:
        del self.jobs

    @property
    def cost(self) -> int:
        """Cost of the sequence, i.e. the maximum completion time
        on all machines

        Returns
        -------
        int
            Cost of the sequence
        """
        return max(ci for c in self.C for ci in c)

    def get_cost(self) -> int:
        """Returns the cost of the sequence, i.e. the maximum completion time
        on all machines

        Returns
        -------
        int
            Cost of the sequence
        """
        return self.cost

    def job_to_bottom(self, job: Job) -> None:
        """Inserts job to the last position of sequence
        (assume sequence is a bottom sequence)

        Parameters
        ----------
        job : Job
            New job included
        """
        self.jobs.append(job)
        # Update on each semiline until reconciliation machine
        c_sl = 0  # The earliest release on the reconciliation machine
        for sl, c in enumerate(self.C):
            c[0] = max(c[0], job.r[sl][0]) + job.p[sl][0]
            for k in range(1, self.m[sl] - job.s):
                c[k] = max(c[k], c[k - 1]) + job.p[sl][k]
            c_sl = max(c[k], c_sl)

        # Update on reconciliation machine
        for sl, c in enumerate(self.C):
            c[-job.s] = max(c_sl, job.r[sl][-job.s]) + job.p[sl][-job.s]
            for k in range(self.m[sl] - job.s + 1, self.m[sl]):
                c[k] = max(c[k], c[k - 1]) + job.p[sl][k]

    def job_to_top(self, job: Job) -> None:
        """Inserts job to the first position of sequence
        (assume sequence is a top sequence)

        Parameters
        ----------
        job : Job
            New job included
        """
        self.jobs.insert(0, job)
        # Update on each semiline on reconciliation machines
        c_sl = 0  # The earliest wait time on the reconciliation machine
        for sl, c in enumerate(self.C):
            m_sl = self.m[sl] - 1
            if m_sl == -1:
                continue
            c[m_sl] = max(c[m_sl], job.q[sl][m_sl]) + job.p[sl][m_sl]
            for _ in range(1, job.s):
                m_sl -= 1
                c[m_sl] = max(c[m_sl], c[m_sl + 1]) + job.p[sl][m_sl]
            c_sl = max(c[m_sl], c_sl)

        # Update on parallel semilines
        for sl, c in enumerate(self.C):
            m_sl = self.m[sl] - job.s - 1
            if m_sl == -1:
                continue
            c[m_sl] = max(c_sl, job.q[sl][m_sl]) + job.p[sl][m_sl]
            for _ in range(1, self.m[sl] - job.s):
                m_sl -= 1
                c[m_sl] = max(c[m_sl], c[m_sl + 1]) + job.p[sl][m_sl]

    def copy(self, deep: bool = False) -> 'Sigma':
        if deep:
            return copy.deepcopy(self)
        sigma = Sigma.__new__(Sigma)
        sigma.jobs = self.jobs.copy()
        sigma.C = [c.copy() for c in self.C]
        sigma.m = self.m
        return sigma

    @staticmethod
    def empty_sigma(m: list[int]) -> 'Sigma':
        """Creates an empty `Sigma` instance of `m` machines on
        each parallel semiline

        Parameters
        ----------
        m : list[int]
            Number of machines considered on each parallel semiline

        Returns
        -------
        Sigma
            New sequence
        """
        return empty_sigma(m)


def empty_sigma(m: list[int]) -> Sigma:
    """Creates an empty `Sigma` instance of `m` machines on
    each parallel semiline

    Parameters
    ----------
    m : list[int]
        Number of machines considered on each parallel semiline

    Returns
    -------
    Sigma
        New sequence
    """
    sigma = Sigma.__new__(Sigma)
    sigma.jobs = []
    sigma.C = [[0] * mi for mi in m]
    sigma.m = m
    return sigma

File: src/bnbprob/slpfssp/cpp/sigma.cpp
#include "sigma.hpp"
#include <algorithm>
#include <limits>
#include <memory>
#include <stdexcept>


// Cost: maximum completion time on all machines
int Sigma::cost() const {
    int max_c = 0;
    for (const auto& c : C) {
        for (int ci : c) {
            if (ci > max_c) max_c = ci;
        }
    }
    return max_c;
}

// Insert job at the end (bottom) of the sequence
void Sigma::job_to_bottom(const JobPtr& job) {
    jobs.push_back(job);
    int c_sl = 0; // The earliest release on the reconciliation machine
    Int2D& p_ = *job->p; // Processing times

    // Update on each semiline until reconciliation machine
    for (size_t sl = 0; sl < C.size(); ++sl) {
        auto& c = C[sl];
        int msl = m->at(sl);
        // Defensive: check sizes
        if (c.empty() || job->r.size() <= sl || p_.size() <= sl) continue;

        c[0] = std::max(c[0], job->r[sl][0]) + p_[sl][0];
        for (int k = 1; k < msl - job->s; ++k) {
            c[k] = std::max(c[k], c[k - 1]) + p_[sl][k];
        }
        c_sl = std::max(c[msl - job->s - 1], c_sl);
    }

    // Update on reconciliation machine
    for (size_t sl = 0; sl < C.size(); ++sl) {
        auto& c = C[sl];
        int msl = m->at(sl);
        if (c.empty() || job->r.size() <= sl || p_.size() <= sl) continue;

        int recon_idx = msl - job->s;
        c[recon_idx] = std::max(c_sl, job->r[sl][recon_idx]) + p_[sl][recon_idx];
        for (int k = recon_idx + 1; k < msl; ++k) {
            c[k] = std::max(c[k], c[k - 1]) + p_[sl][k];
        }
    }
}

// Insert job at the beginning (top) of the sequence
void Sigma::job_to_top(const JobPtr& job) {
    jobs.insert(jobs.begin(), job);
    int c_sl = 0; // The earliest wait time on the reconciliation machine
    Int2D& p_ = *job->p; // Processing times

    // Update on each semiline on reconciliation machines (backwards)
    for (size_t sl = 0; sl < C.size(); ++sl) {
        auto& c = C[sl];
        int msl = m->at(sl);
        if (c.empty() || job->q.size() <= sl || p_.size() <= sl) continue;

        int idx = msl - 1;
        if (idx < 0) continue;
        c[idx] = std::max(c[idx], job->q[sl][idx]) + p_[sl][idx];
        for (int s = 1; s < job->s; ++s) {
            --idx;
            if (idx < 0) break;
            c[idx] = std::max(c[idx], c[idx + 1]) + p_[sl][idx];
        }
        c_sl = std::max(c[idx], c_sl);
    }

    // Update on parallel semilines (backwards)
    for (size_t sl = 0; sl < C.size(); ++sl) {
        auto& c = C[sl];
        int msl = m->at(sl);
        if (c.empty() || job->q.size() <= sl || p_.size() <= sl) continue;

        int idx = msl - job->s - 1;
        if (idx < 0) continue;
        c[idx] = std::max(c_sl, job->q[sl][idx]) + p_[sl][idx];
        for (int s = 1; s < msl - job->s; ++s) {
            --idx;
            if (idx < 0) break;
            c[idx] = std::max(c[idx], c[idx + 1]) + p_[sl][idx];
        }
    }
}

// Deepcopy of self
Sigma Sigma::deepcopy() const {
    Sigma other = Sigma();
    other.m = this->m;
    other.C = this->C;
    other.jobs = copy_jobs(this->jobs);
    return other;
}
